rust/otap-dataflow/crates/otap/src/otap_grpc/otlp/server.rs

```diff
@@ -145,15 +199,61 @@ impl UnaryService<OtapPdata> for OtapBatchService {
     type Future = BoxFuture<'static, Result<tonic::Response<Self::Response>, Status>>;
 
     fn call(&mut self, request: tonic::Request<OtapPdata>) -> Self::Future {
-        let otap_batch = request.into_inner();
+        let mut otap_batch = request.into_inner();
 
         let effect_handler = self.effect_handler.clone();
+        let state = self.state.clone();
         Box::pin(async move {
+            // Allocate slot with closure that creates the channel only if slot is available
+            let (slot_key, rx) = {
+                let mut state = state.lock().expect("lock acquired");
+                match state.allocate_slot(|| oneshot::channel()) {
+                    Some((key, rx)) => (key, rx),
+                    None => {
+                        return Err(Status::resource_exhausted("Too many concurrent requests"));
+                    }
+                }
+            };
+
+            // Create guard to clean up slot if this future is cancelled/dropped
+            struct SlotGuard {
+                slot_key: SlotKey,
+                state: SharedCorrelationState,
+            }
+
+            impl Drop for SlotGuard {
+                fn drop(&mut self) {
+                    if let Ok(mut state) = self.state.lock() {
+                        state.cancel(self.slot_key);
+                    }
+                }
+            }
```

Comment:
Why having this struct declaration inside a future?

Reply:
Fixed, cleaned up!